Analogue audio ala Frodo

Aldzi opened the pull request:
XBMC Frodo and earlier on Raspberry Pi being controlled from UPnP controllers from a smartphone allows playing music and changing volume directly from controller. After Frodo user has to set audio to maximum volume to get an audible sound from the analogue output. This is done because analogue audio output has limited dynamic range of about 11 bits. Users are advised changing volume on amplifier.

As result of this choice users have to select music and then search for remote from the amplifier or walk to it to change the volume. It is rather awkward. This simple action could be difficult to fulfil if amplifier is hidden somewhere in the room away from the sight.

This behavior also dramatically different from behavior of other DLNA renderers, where the user can directly change the volume from a controller.

This pull request provides an option to handle analogue audio output as in Frodo and before. This option is available from System Settings > System >Audio Output > Use analogue amplifier. This option is switched off by default.

Aldzi committed:
Implementation of optional function to handle analogue output on Raspberry Pi as it was in XBMC Frodo and before. After Frodo, amplifier wasn't used anymore

anaconda commented on test.txt:
```diff
@@ -0,0 +1 @@
+Here comes the sun!
```
Oops?

Aldzi replied:
Hm... I thought that I removed this file... It was trial with my repository. I have quite little understanding of  how it works. I'm new to it.

Do you have comments about other changes?

Razzeee commented on system/settings/rbp.xml:
```diff
@@ -72,6 +72,16 @@
+          <level>0</level>
+          <default>false</default>
+          <control type="toggle" />
+          <updates>
```
I don't think you need this, as there is nothing to update, because this setting is new.

Aldzi replied:
This change allow changing "use amplifier" option in UI. What exactly do you mean?

da-anda replied:
The `<update>` part/section might have been needed for the setting you used as reference but it's not necessary for your new setting, as there is no old value to migrate/update. So just remove this section.

Aldzi replied:
Thank you!

Razzeee commented:
I don't really like platform specific stuff, but it's not my decision.

Ping @popcornmix maybe @fritsch and @FernetMenta too?

fritsch commented:
That code is not clean - the bits in AESinkPI.cpp work and the menu integration most likely works, too - the ActiveAE changes are wrong.

@popcornmix's code mostly.

popcornmix commented:
Not really sure why this needs a GUI setting and an advancedsetting. I would have thought either treat a non-default advancedsetting as an enable and drop the GUI setting. Or make the GUI handle the amount of amplification.

Second point is when playing video, does the "Amplification" OSD option not provide the behaviour you want? E.g. set amplification to 12dB, then you should be able to reduce the analogue audio's volume from remote without wasting bits of dynamic range.

I know that doesn't solve the problem with music, but may suggest other ways of achieving this goal.

Aldzi commented:
@fritsch - I added just one line in CActiveAE::OnSettingsChange intended to detect that the settings are changed: "+      setting == CSettings::SETTING_AUDIOOUTPUT_USEANALOGAMPLIFIER     ||". The rest of the changes are generated by git - bringing my version in to sync with the one in xbmc. Could you please to be more specific on what you see as wrong.

fritsch commented:
You cannot just "sync" something ... you remove code that we need!

Aldzi commented:
@fritsch - exactly not. I did not remove those lines: "format.m_sampleRate = 192000;" & etc.. They are present in my branch. It is git who says that those lines will be removed. And it is also correct - they are absent in the master.

fritsch commented:
I neither want: "Here comes the sun" nor "It was a trial. Must go away..."  https://patch-diff.githubusercontent.com/raw/xbmc/xbmc/pull/8704.patch in any kodi code - now fix your stuff, please.

Aldzi commented:
@fritsch - "here comes the sun" is already removed. I'm surprised you still see it. How can I remove it? Sorry, I'm new to it...

fritsch commented:
rebase your branch onto kodi's master branch - only pick the patches you want and fix the one I pointed out now twice - then force push to this branch.

Aldzi commented:
@fritsch - I barely understand what you say. I'll do my best.

Razzeee commented:
@Aldzi 
you might want to read https://help.github.com/articles/using-git-rebase/ and before you start with that, make a backup of your folder. Just in case, helps when you start out and learn GIT.

Aldzi commented:
@popcornmix - I do not understand why you say about "why this needs a GUI setting and an advancedsetting". As far as I see I did nothing with advanced setting but GUI only. Could you pinpoint for me the changes in advancedseetings. If there are changes - thay aren't intended.

About the second point - when user plays video, user sees display and all of the options are available. User is able to do whatever he wants. Also, it's very possible that he uses pass through to the TV/amplifier/receiver via HDMI. In such case, this option is not relevant all. That's why it's off by default.

I'm solving the problem when a) the amplifier is not reachable or difficult to reach b) make Kodi on RPi and an amplifier combination behave similarly as any other DLNA renderer. In such case user doesn't need GUI and, probably, it is not available.

Aldzi commented:
@Razzeee - thank you!

popcornmix commented:
You are using g_advancedSettings.m_ac3Gain as the amount to boost the volume by. See:
https://github.com/Aldzi/xbmc/commit/c1092c131ff6dcc8f8acc0674dd84d59503ab077#diff-c7c594453ceb6053fcaf9bbedd42ef79R523

MartijnKaijser commented:
Before there's any more effort put in this are we gonna except it or not? Personally I'd say no

NedScott commented:
It would be a good idea to be encouraging to someone who's willing to make PRs. Even if it doesn't get added, shooting things down right away just discourages new involvement.

Aldzi commented:
@popcornmix - ok, I see g_advancedSettings.m_ac3Gain. I put it in because it is also used in some other code. Do you want me remove it? I have no problem.

What does it mean "except"?

MartijnKaijser commented:
Why? I just wanted to it clear that either it goes it or not. Saves possible wasted effort. Doesn't mean any other possible involvement will be refused. There are already to many PRs open that will never go in or stay in limbo.

Aldzi commented:
What's wrong with the proposal?

fritsch commented:
Here is what is wrong with the proposal.

What we have currently:
Basically the code uses an advancedsetting member (ac3) from something totall unrelated, additionally it wants a gui setting, touches global ActiveAE engine code while polluting every platform with something RPi specific.

We have two additional patches that are in no way related.

The code bases on some other branch / code and won't apply as is (really wondering why github is green).

To fix those issues:

- rebase
- remove the commits we have no interest
- don't reuse other advancedsetting variables
- don't spam the ActiveAE main for RPI specifics

When this is done @popcornmix can decide if he wants it like it is.

Aldzi commented:
I have no problem with the fixes.

What's the process? After those changes made - do I need to supply another pull request or what?

popcornmix commented:
To be honest I'm not really understanding the issue this PR is designed to solve.

Analogue audio volume can be controlled in just the same way as hdmi volume. The only issue is the limited dynamic range of analogue audio means reducing the volume harms quality and some "amplification" may be desirable.

The lack of an amplification option in paplayer could be an issue. But I'm not sure this is the issue @Aldzi has. Possibly it will be more obvious when the PR is rebased and just the desired changes are present.

Razzeee commented:
@Aldzi 
just rebase it and force push it, as mentioned in the link above. when you force push, github will know what you did and update this PR, so you won't need to do another one.

Aldzi commented:
@Razzeee - thanks, will do.

Aldzi commented:
Hi there,

I succeed to:
* rebase
* removed the commits of no interest
* removed usage of advancedsetting variables

Unfortuantely, I did not suceed to revert changes to ActiveAE. How can I do it?

Razzeee commented:
Change the file back to it's previous state, make a new commit, the message doesn't matter. Rebase onto master again. (if you must force rebase)
Then select squash on the last commit, it will be merged with the first one.

Aldzi commented:
@Razzeee - the file is already at its original state. There are no my changes. Anyways, thanks, will try.

Aldzi commented:
@Razzeee - thank you for your support!

Aldzi commented:
@popcornmix - the change is rebased and contains only necessary modifications. Your comments, please.

popcornmix commented:
I believe the parameter to SetVolume is logarithmic with 0.0=-60dB to 1.0=0dB.
You are treating it as a linear value, which means it will be correct at 0 and 1 but incorrect between.

All OMX_IndexConfigAudioVolume does is scale the submitted samples. I don't see what difference that will have compared to letting the audio engine scale the samples.

Do you believe after this pull request that 100% volume is louder, quieter or the same?
Do you believe after this pull request that 50% volume is louder, quieter or the same?

Aldzi commented:
Indeed, good question.

I just reuse the code as it was in XBMC Frodo and before with minor modification with AC3 gain as you requested. I don't have knowledge of RPi electronics. So, if this code is wrong, then it was also wrong in Frodo. I asked about my intensions: http://forum.kodi.tv/showthread.php?tid=252889, but guess what - I got no answer.

Subjectively, I hear that Frodo code at 100% volume is the same as audio engine (AE) scaling. At 50% volume Frodo is higher then AE scaling. It remains audible even at very low volume being higher then noise level.

As far as I understand, AE uses logarithmic scale like shown below (After Frodo): ![Scales](https://raw.githubusercontent.com/Aldzi/various/master/kodi/2ch%20volume%20graphs.png) Probably, scale is event stronger then logarithmic, because at low levels sound is lost in RPi noise.

I see that I propose a linear scale of volume handling instead of logarithmic. I still see it usefull. Please, your reaction.

popcornmix commented:
Back on Frodi, The Pi used omxplayer for everything (video and music).
Since Gotham we started using paplayer for music and on Helix we use dvdplayer for video (on Pi 2).

Also back on Frodo, omxplayer had a bug which was fixed here: https://github.com/xbmc/xbmc/pull/6576
This PR seems to be reintroducing that bug for paplayer and dvdplayer.

I'm afraid that is not going to happen. The current behaviour is consistent between omxplayer and paplayer/dvdplayer and matches the behaviour of all other kodi platforms.

Aldzi commented:
I compared volume control graphs of Kodi on RPi B+ and Denon AVR-X1000 amplifier. I see that both use logarithmic curve.

Indeed, I would not waste a lot of my time if the reference to https://github.com/xbmc/xbmc/pull/6576 was given earlier or intensions were discused at: http://forum.kodi.tv/showthread.php?tid=252889

Thanks
